Maker page sub nav, App directory page styles

kyranjamie opened the pull request:
This PR introduces the design changes for the [navigation menu](https://www.figma.com/file/aXnNi8MOmN3CDrVwaDXXuqX3/Product?node-id=3360%3A1279).

- Active nav menu
- ESLinting TypeScript
- WIP select list
- App directory page `/maker/apps`
- Updates to app submit page

kyranjamie commented on components/top-bar/index.tsx:
```diff
@@ -0,0 +1,86 @@
+      { admin && generateLinkList(adminLinks)({ isErrorPage }) }
+      { generateLinkList(primaryNavLinks)({ isErrorPage }) }
+    </>
+  )
```
Curious what the pattern to follow is here. I _could_ create a totally separate component, but then I have to repeat the props. 
What's the best way to keep this clean and not repeat the contents?

kyranjamie commented on components/maker/nav/maker-nav.tsx:
```diff
@@ -0,0 +1,96 @@
+import React from 'react'
```
This component is still WIP

hstove commented:
A few things blocking me from reviewing:

- Couldn't find `@blockstack/ui` (easily fixed locally)
- Type error in `/navigation/helpers.tsx` (couldn't figure out a quick fix here)
- ESLint errors on `tsx` pages (not linting errors, just that we haven't setup typescript yet with eslint). Let's make sure we have `@typescript-eslint` setup

hstove commented:
Yeah, I can look into that, I think it just requires a few tweaks to `.eslintrc`

hstove commented:
Is there somewhere that you've described exactly what issues you're having with redux? You mentioned it's overwriting the store on page load, but I can't tell exactly where those changes were made to debug the issue. If you can point me in the right direction, I'll take another look.

Is this related to the introduction of the `jwt` in cookies? If so, I would recommend falling back to the previous approach that was working. Obviously SSR is better, but not worth it if it's causing this much of a delay.

hstove commented:
> Added eslint for TypeScript, mostly works but having this error when importing an exported interface

I was able to fix that one in #668

hstove commented:
Overall, I was able to review the nav changes, app listing page, and ESLint stuff, and that looks good! Would just like to see this branch in a fully buildable state (i.e. heroku deploy works) before merging.

kyranjamie commented:
Thanks for making the PR @hstove 

The best example to demonstrate the issue is on the `/maker/apps/:appId` route. When it loads SSR'd, the app seems mostly hydrated as expected, but examining the store client side, you can see `state.maker` is in its initial state. Logging the output of the `getInitialProps` shows that it is populating the store correctly server side.

Happy to switch back to fetching this data client side in an effect, but this isn't really fixing the issue. Plus, the drive to do it this way came from difficulties implementing it client side in cases where the data isn't there on initial render.

aulneau commented:
> Is there somewhere that you've described exactly what issues you're having with redux? You mentioned it's overwriting the store on page load, but I can't tell exactly where those changes were made to debug the issue. If you can point me in the right direction, I'll take another look.

> Is this related to the introduction of the jwt in cookies? If so, I would recommend falling back to the previous approach that was working. Obviously SSR is better, but not worth it if it's causing this much of a delay.

I would be surprised if the redux issues came from the implementation of the cookie for jwt. I think a refactoring of our redux implementation to update it to the latest version will probably result it in it working correctly (also updating it based off of the next.js examples for `with-redux`).

hstove commented on components/maker/empty-state/no-apps.tsx:
```diff
@@ -0,0 +1,26 @@
+      <Text mb={6}>
+        Submit your app or sign in with a different Blockstack ID.
+      </Text>
+      <Button onClick={() => router.push('/submit-your-app')}>
```
Similar story here with accessibility. It's better to use a `<Link passHref ...>` and `<Button is="a">`. If the new UI library doesn't support this well, we can punt on it for now, but we would need to track that as an issue in `@blockstack/ui`.

kyranjamie replied:
Button loses styling when using `is="a"`, to investigate

hstove commented on components/maker/payment/index.js:
```diff
@@ -21,8 +21,10 @@ const PaymentDetails = ({ app, apiServer, accessToken, user, dispatch }) => {
   const isSaved = (
```
I think you can reduce the `addr !== '' && addr !== null` with just `!!addr`.

hstove commented on components/navigation/helpers.tsx:
```diff
@@ -0,0 +1,48 @@
+import { StyledAnchor } from '@components/top-bar/styled'
+import { NavLink } from '@common/constants'
+
+export const ErrorPageLink: React.FC<any> = ({ isErrorPage, children, href, ...props }) => (
```
Nit: Can we make an explicit interface for these props? It would be great to not have explicity `any`s as we migrate to TS.

hstove commented on components/navigation/index.tsx:
```diff
@@ -0,0 +1,20 @@
+import NewsletterActions from '@stores/newsletter/actions'
+import { StyledTopBar } from '@components/top-bar/styled'
+
+const NavigationComponent: React.FC<any> = ({ children, mobile, display, variant }) => {
```
Another case of no explicit `any` - this one should be easy to define.

hstove commented on components/page/index.tsx:
```diff
@@ -0,0 +1,41 @@
+  innerPadding?: number | number[];
+  subNav?: React.ReactNode;
+  align?: string;
+  [prop: string]: any;
```
I'm still figuring out this situation with TS & React - do we need to have this line? Can TS be smart about the fact that you're passing props to a child?

kyranjamie replied:
It is indeed needed, as we're saying: this component accepts these properties... and is used as a dictionary (of other style props). TypeError if removed.

The UI library _could_ expose some types to extend (that include all the base styled system props), in order to type it explicitly. However, I think this raises more of a question about how we build components. Ideally, imo, we don't want to mix the styling of components too much with the data/functionality they accept, otherwise important props, `admin={true/false}`get lost amongst `borderTop="grey"`.

An exception to this may be style variants a component needs, but even think it's better to encapsulate with true `variant` props, rather than exposing the entire styled-system suite of props. Maybe this goes against the whole paradigm of styled-system though? But definitely think there's merit it decoupling styles to some degree.

hstove commented on components/submit/blockstack-id-card.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
```
Is this component going to be used? I don't see it anywhere yet - I'd assume maybe it'll end up in `MakerNav`

kyranjamie replied:
used on `submit-your-app.tsx` page

hstove commented on components/maker/nav/maker-nav.tsx:
```diff
@@ -0,0 +1,92 @@
+import { Arrow } from '@components/arrow'
+import { MakerNavContainer } from './nav-layout'
+
+const StyledSelectBox = styled(Box)`
```
Oh, also, this specific component isn't being used anywhere (yet?)

hstove commented on models/app.ts:
```diff
@@ -0,0 +1,41 @@
+export interface App {
```
nice 👍

kyranjamie replied:
Best part about using TypeScript across the stack, at some point you can share models between the backend/frontend

hstove commented on pages/_app.js:
```diff
@@ -73,6 +88,21 @@ const GlobalStyles = createGlobalStyle`
 
+const REFACTORED_PATHS = ['/maker/apps', '/maker/apps/blockstack-only', '/submit-your-app'];
+
+const RenderRouteThemeProvider = ({ children }) => {
```
Nice! Interesting way to gradually introduce new styles.

kyranjamie replied:
It isn't perfect, still some weird behaviours, but a step in the right direction.

hstove commented on pages/maker/apps/[appId].tsx:
```diff
@@ -0,0 +1,137 @@
+          <Box>
+            <MakerContentBox>
+              <Maker.PaymentDetails
+                app={selectedApp}
```
Nit: I know you're just re-using the code I originally wrote here, but it would probably be cleaner to just select these props from redux inside of the component.

hstove commented on pages/maker/apps/index.tsx:
```diff
@@ -0,0 +1,75 @@
+}
+
+const AppDirectoryPageContainer: React.FC = ({ children }) => (
+  <ThemeProvider theme={theme}>
```
Do we need this? Shouldn't we get the same new theme from `_app.js`?

hstove commented on stores/user/index.ts:
```diff
@@ -1,104 +1,106 @@
 
-const handleSignIn = (apiServer) => async (dispatch) => {
-  const token = userSession.getAuthResponseToken()
+const handleSignIn = (apiServer: any) => async (dispatch: any) => {
```
`apiServer` is obviously `string`, and you can get `import { Dispatch } from 'redux'` for the `dispatch` type.

kyranjamie commented on components/list/apps/list.js:
```diff
@@ -75,7 +75,7 @@ class AppsListComponent extends React.Component {
   }
 
-  componentWillReceiveProps(nextProps) {
+  UNSAFE_componentWillReceiveProps(nextProps) {
```
Yeah this is part of their deprecation process. To be refactored, but for the time being they recommend a rename.

kyranjamie commented on components/app-directory/index.tsx:
```diff
@@ -0,0 +1,55 @@
+
+  return (
+    <Flex>
+      <Link href={url}>
```
The whole area is clickable. Height etc is defined on a child of `<Link />`

kyranjamie commented on components/app-directory/index.tsx:
```diff
@@ -0,0 +1,55 @@
+  const url =
+    app.authentication === 'Blockstack'
+      ? `/maker/apps/${app.id}`
+      : '/maker/blockstack-only';
```
Handling server side now.

kyranjamie commented on pages/maker/apps/[appId].tsx:
```diff
@@ -0,0 +1,137 @@
+  const subNav = (
+    <MakerNav
+      apps={appList}
+      userId="kyranjamie.id"
```
Updated

kyranjamie commented on components/arrow/index.tsx:
```diff
@@ -0,0 +1,30 @@
+import React from 'react'
```
Trying to format/lint as I go, but know these'll keep cropping up until I get CI in place to auto-tell me off.

kyranjamie commented:
Thanks for the review @hstove, addressed most of these points now.

Curious for your input regarding the styling issues/glitches that happen, primarily in dev mode. I'm almost certain it's to do with the styled-components warning that there are multiple instances running.

Presently, especially on the submit page, every reload renders twice, first in one style, then again in another.

kyranjamie commented:
Now QA has started, merging this into release branch to work from,
